[v1][Metrics] Add design doc

markmc opened the pull request:
Related to #10582. Some notes I had taken on v0 metrics implementation, along with v1 design details.

markmc committed:
[WIP][v1][Metrics] Add design doc

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

MikeSpreitzer commented:
FYI, this use of colons in direct metric names is contrary to https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels

I see that most follow the desideratum of ending with the units, but `time_in_queue_requests` does not.

While backward compatibility is good, maybe there should also be a plan to shift to more compliant metric names?

hmellor commented:
When the metrics were originally added the contributor wasn't aware of the naming convention. When this was noticed it was decided that we would leave them as is so that we wouldn't break any dashboards that people had set up. 

I agree that this could be a good opportunity to switch to compliant names. I think using the correct convention would inspire confidence in users who are already familliar with tools like Prometheus.

russellb commented on docs/source/design/v1/metrics.md:
```diff
@@ -0,0 +1,354 @@
+- `process_open_fds`
+- `process_max_fds`
+
+(Relevant because if we move away from multiprocess mode in v1, we get these back)
```
I'm not sure how useful they are, though, because I assume they only represent the API server process and not the multiple underneath.

hmellor commented on docs/source/design/v1/metrics.md:
```diff
@@ -0,0 +1,354 @@
+- The number of new tokens generated per second over the past 5
+  seconds
+
+## Metrics Publishing - Prometheus
```
Should this be `###` like the rest?

markmc replied:
Thanks, think I've fixed that now

hmellor reviewed (approved):
LGTM, just some nits about cross referencing syntax and the heading level of one of the titles

markmc commented on docs/source/design/v1/metrics.md:
```diff
@@ -0,0 +1,354 @@
+# Problem Statement
```
Thanks!

markmc commented on docs/source/design/v1/metrics.md:
```diff
@@ -0,0 +1,354 @@
+- `vllm:request_decode_time_seconds` - Requests Decode Time
+- `vllm:request_max_num_generation_tokens` - Max Generation Token in Sequence Group
+
+See [the PR which added this Dashboard](https://github.com/vllm-project/vllm/pull/2316) for interesting and useful background on the choices made here.
```
I'd be delighted if you had the time to do those fixups? I'll try to come back to it if not :+1:

markmc commented:
> When the metrics were originally added the contributor wasn't aware of the naming convention. When this was noticed it was decided that we would leave them as is so that we wouldn't break any dashboards that people had set up.
> 
> I agree that this could be a good opportunity to switch to compliant names. I think using the correct convention would inspire confidence in users who are already familliar with tools like Prometheus.

I added a section on this, thanks :+1:

hmellor commented on docs/source/design/v1/metrics.md:
```diff
@@ -0,0 +1,354 @@
+- `vllm:spec_decode_num_draft_tokens_total` (Counter)
+- `vllm:spec_decode_num_emitted_tokens_total` (Counter)
+
+These are documented under [Inferencing and Serving -> Production Metrics](https://docs.vllm.ai/en/stable/serving/metrics.html).
```
I'd be happy to, but I don't have maintainer permissions so I'm not able to make edits to your PR

russellb reviewed (approved):
lgtm! Thank you for covering this so thoroughly! It should save others a ton of time that come behind you and need to catch on up the state of metrics support.

I also really appreciate the guidance on how to deprecate metrics, as well as the opportunities for future work.

hmellor commented on docs/source/design/v1/metrics.md:
```diff
@@ -0,0 +1,713 @@
+# HELP vllm:cache_config_info Information of the LLMEngine CacheConfig
+# TYPE vllm:cache_config_info gauge
+vllm:cache_config_info{block_size="16",cache_dtype="auto",calculate_kv_scales="False",cpu_offload_gb="0",enable_prefix_caching="False",gpu_memory_utilization="0.9",...} 1.0
+
```
```suggestion
```

hmellor commented on docs/source/design/v1/metrics.md:
````diff
@@ -0,0 +1,713 @@
+Put another way:
+
+```text
+  << queued timestamp >>
+    [ queue interval ]
+      |
+      | (possible preemptions)
+      | << scheduled timestamp >>
+      | << preempted timestamp >>
+      | << scheduled timestamp >>
+      | << new token timestamp (FIRST) >>
+      | << new token timestamp >>
+      | << new token timestamp >>
+      | << preempted timestamp >>
+      v
+  << scheduled timestamp >>
+    [ prefill interval ]
+  << new token timestamp (FIRST) >>
+    [ inter-token interval ]
+  << new token timestamp >>
+    [ decode interval (relative to most recent first token time)
+    [ inference interval (relative to most recent scheduled time)
+  << new token timestamp (FINISHED) >>
````
```suggestion
<< queued timestamp >>
  [ queue interval ]
    |
    | (possible preemptions)
    | << scheduled timestamp >>
    | << preempted timestamp >>
    | << scheduled timestamp >>
    | << new token timestamp (FIRST) >>
    | << new token timestamp >>
    | << new token timestamp >>
    | << preempted timestamp >>
    v
<< scheduled timestamp >>
  [ prefill interval ]
<< new token timestamp (FIRST) >>
  [ inter-token interval ]
<< new token timestamp >>
  [ decode interval (relative to most recent first token time)
  [ inference interval (relative to most recent scheduled time)
<< new token timestamp (FINISHED) >>
```

hmellor reviewed (commented):
This is a really valuable document, thank you for putting the time and effort into creating it!

A couple of whitespace nits. 

My comments about cross-referencing still stand. I'd do it myself, but I don't have GitHub permission to.

hmellor committed:
Use myst links where possible

Signed-off-by: Harry Mellor <19981378+hmellor@users.noreply.github.com>

hmellor commented:
You can cherrypick this commit https://github.com/hmellor/vllm/commit/1876f9cc5123f74239779d38005dc80f0c7552c3

markmc commented:
> You can cherrypick this commit [hmellor@1876f9c](https://github.com/hmellor/vllm/commit/1876f9cc5123f74239779d38005dc80f0c7552c3)

Thanks @hmellor appreciate the help!

TIL MyST!

markmc committed:
[v1][Metrics] Add interval calculation diagrams

As per the discussion in #13169.

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

hmellor reviewed (approved):
Thanks for cherry picking the links fix, my comments have been resolved

robertgshaw2-redhat commented:
It seems like there might be an actual issue with the docs build

comaniac commented on docs/source/design/v1/metrics.md:
```diff
@@ -0,0 +1,711 @@
+is possible), we consider the preemption as affecting the
+time-to-first-token and prefill intervals.
+
+:::{image} /assets/design/v1/metrics/intervals-2.png
```
`intervals-3`?

markmc replied:
Thanks!

comaniac commented on docs/source/design/v1/metrics.md:
```diff
@@ -0,0 +1,711 @@
+
+### Metrics Publishing - Logging
+
+The `LoggingStatLogger` metrics publisher outputs a log `INFO` message
+every 5 seconds with some key metrics:
```
Better to mention the prefix caching metric here given it's first class citizen and the semantic in logging is different from v0. In v0, the reported hit rate is accumulated from server launch. In v1, the reported hit rate is based on the most recent 1k block queries.

markmc replied:
Good point, that came after I wrote this doc :+1:

comaniac commented on docs/source/design/v1/metrics.md:
```diff
@@ -0,0 +1,711 @@
+
+In the case of logging, we expect the user is best served by
+calculating the hit rate over a fixed number of the most recent
+queries.
```
```suggestion
queries (the interval is fixed to 1k most recent queries for now).
```

markmc committed:
[v1][Metrics] Fix diagram link

Pointed out by Cody.

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc committed:
[V1][Metrics] Mention prefix cache metric in logging logger

Pointed out by Cody.

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

comaniac commented on docs/source/design/v1/metrics.md:
```diff
@@ -0,0 +1,712 @@
+  seconds
+- The number of new tokens generated per second over the past 5
+  seconds
+- The prefix cache hit rate over the most recent 1 queries
```
```suggestion
- The prefix cache hit rate over the most recent 1k kv-cache block queries
```

comaniac reviewed (approved):
LGTM. Thanks

markmc committed:
[V1][Metrics] Add note on prefix cache metric

Co-authored-by: Cody Yu <hao.yu.cody@gmail.com>
Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc committed:
[V1][Metrics] Tweak prefix caching note

Co-authored-by: Cody Yu <hao.yu.cody@gmail.com>
Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc committed:
[V1][Metrics] Fix docs build in buildkite

The MyST examples links are causing this in the buildkite build:

```
docs/source/design/v1/metrics.md:64:Unknown source document 'getting_started/examples/prometheus_grafana' [myst.xref_missing]
```

Signed-off-by: Mark McLoughlin <markmc@redhat.com>
